winml/test/adapter/AdapterSessionTest.cpp

```diff
@@ -0,0 +1,311 @@
+  auto session = CreateUniqueOrtSession(session_options);
+  THROW_IF_NOT_OK_MSG(winml_adapter_api->SessionRegisterCustomRegistry(session.get(), registry), ort_api);
+}
+
+void LoadAndPurloinModel(const UniqueOrtSession& session, const std::string& model_path) {
+  com_ptr<WinML::IModel> model;
+  WINML_THROW_IF_FAILED(engine_factory->CreateModel(model_path.c_str(), sizeof(model_path), model.put()));
+
+  Microsoft::WRL::ComPtr<WinML::IOnnxruntimeModel> onnxruntime_model;
+  WINML_EXPECT_HRESULT_SUCCEEDED(model->QueryInterface(IID_PPV_ARGS(&onnxruntime_model)));
+  OrtModel* ort_model;
+  WINML_EXPECT_HRESULT_SUCCEEDED(onnxruntime_model->DetachOrtModel(&ort_model));
+  THROW_IF_NOT_OK_MSG(winml_adapter_api->SessionLoadAndPurloinModel(session.get(), ort_model), ort_api);
+}
+
+void LoadAndPurloinModel() {
+  const auto session_options = CreateUniqueOrtSessionOptions();
+  auto session = CreateUniqueOrtSession(session_options);
+  LoadAndPurloinModel(session, "fns-candy.onnx");
+}
+
+void Initialize() {
+  const auto session_options = CreateUniqueOrtSessionOptions();
+  auto session = CreateUniqueOrtSession(session_options);
+
+  com_ptr<WinML::IModel> model;
+  const auto model_path = "fns-candy.onnx";
+  WINML_THROW_IF_FAILED(engine_factory->CreateModel(model_path, sizeof(model_path), model.put()));
+
+  Microsoft::WRL::ComPtr<WinML::IOnnxruntimeModel> onnxruntime_model;
+  WINML_EXPECT_HRESULT_SUCCEEDED(model->QueryInterface(IID_PPV_ARGS(&onnxruntime_model)));
+  OrtModel* ort_model;
+  WINML_EXPECT_HRESULT_SUCCEEDED(onnxruntime_model->DetachOrtModel(&ort_model));
+  THROW_IF_NOT_OK_MSG(winml_adapter_api->SessionLoadAndPurloinModel(session.get(), ort_model), ort_api);
+
+  THROW_IF_NOT_OK_MSG(winml_adapter_api->SessionInitialize(session.get()), ort_api);
+}
+
+class MockSink : public onnxruntime::logging::ISink {
+public:
+  static bool called;
+
+private:
+  void SendImpl(const onnxruntime::logging::Timestamp& timestamp, const std::string& logger_id, const onnxruntime::logging::Capture& message) override {
+    called = true;
+  }
+
+  void SendProfileEvent(onnxruntime::profiling::EventRecord&) const override { };
+};
+bool MockSink::called = false;
+
+void Profiling() {
+  ort_env->SetLoggingManager(nullptr);
+  const std::string logger_name("Default");
+  auto logging_manager = std::make_unique<onnxruntime::logging::LoggingManager>(
+    std::make_unique<MockSink>(),
+    onnxruntime::logging::Severity::kINFO,
+    false,
+    onnxruntime::logging::LoggingManager::InstanceType::Default,
+    &logger_name);
+  ort_env->SetLoggingManager(std::move(logging_manager));
+
+  const auto session_options = CreateUniqueOrtSessionOptions();
+  auto session = CreateUniqueOrtSession(session_options);
+
+  winml_adapter_api->SessionStartProfiling(ort_env, session.get());
+  LoadAndPurloinModel(session, "fns-candy.onnx");
+  winml_adapter_api->SessionEndProfiling(session.get());
+  WINML_EXPECT_TRUE(MockSink::called);
+}
```

Comment:
This test is crashing occasionally (but not always) with a SEGFAULT. The debugger shows some invalid contents in the logging manager. It looks like it maybe got freed while still in use, and may be some race condition managing the logging manager.

Reply:
This is not a thread safe api to call, and it not exposed on the ABI. It is only there so that the WinML API can configure a sink and needs to be called using the same pattern in the ort lib.

Use EnvConfigureCustomLoggerAndProfiler on the winml_adapter_api